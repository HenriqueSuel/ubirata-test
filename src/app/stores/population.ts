import { create } from "zustand";
import { devtools } from "zustand/middleware";

import {
  FeatureCollection,
  PopulationNeighborhoods,
  Properties,
} from "../interfaces/neighborhoodGeometries.interface";
import { getApi } from "../service/api.service";
import { handleSelectFilter } from "../utils/populations.utils";

interface IHandleSelectProps {
  years: string;
  idNeighborhood: number;
}

export interface PopulationState {
  selectedData: {
    years: string;
    idNeighborhood: number | null;
    neighborhood: Properties | null;
    population: PopulationNeighborhoods[];
  };
  filters: {
    neighborhoods: Properties[];
    years: string[];
  };
  neighborhoodGeometries: FeatureCollection | null;
  populationNeighborhoods: PopulationNeighborhoods[];
  handleSelectData: ({
    years,
    idNeighborhood,
  }: Partial<IHandleSelectProps>) => void;
  handleGetData: () => Promise<void>;
}

const usePopulation = create<PopulationState>()(
  devtools((set, get) => ({
    selectedData: {
      years: "",
      idNeighborhood: null,
      neighborhood: null,
      population: [],
    },
    filters: {
      neighborhoods: [],
      years: [],
    },
    neighborhoodGeometries: null,
    populationNeighborhoods: [],
    handleGetData: async () => {
      const [populationNeighborhoods, neighborhoodGeometries] =
        await Promise.all([
          getApi("mocks/populacao_bairros.json") as Promise<
            PopulationNeighborhoods[]
          >,
          getApi("mocks/geometrias_bairros.json") as Promise<FeatureCollection>,
        ]);

      const filters = handleSelectFilter(
        neighborhoodGeometries,
        populationNeighborhoods
      );

      const populationInitial = populationNeighborhoods.filter(
        (population) => population.id_geometria === filters.neighborhoods[0].id
      );

      set(() => ({
        filters,
        selectedData: {
          idNeighborhood: filters.neighborhoods[0].id,
          neighborhood: filters.neighborhoods[0],
          population: populationInitial,
          years: filters.years[0],
        },
        neighborhoodGeometries,
        populationNeighborhoods,
      }));
    },
    handleSelectData: ({ idNeighborhood, years }) => {
      const existingIdNeighborhood =
        idNeighborhood || get().selectedData.idNeighborhood;

      const neighborhood = get().filters.neighborhoods.find(
        (neighborhood) => neighborhood.id === existingIdNeighborhood
      );

      const populationNeighborhoods = get().populationNeighborhoods.filter(
        (item) => item.id_geometria === existingIdNeighborhood
      );

      set((state) => ({
        selectedData: {
          neighborhood: neighborhood || null,
          population: populationNeighborhoods,
          years: years || state.selectedData.years,
          idNeighborhood: existingIdNeighborhood,
        },
      }));
    },
  }))
);

export { usePopulation };
