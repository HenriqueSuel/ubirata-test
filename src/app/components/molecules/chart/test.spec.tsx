import { render, screen } from '@testing-library/react';
import { usePopulation } from '@/app/stores/population';
import { Chart } from './';
import { Mock } from 'vitest';

vi.mock('@/app/stores/population', () => ({
    usePopulation: vi.fn(),
}));

vi.mock("recharts", async () => {
    const OriginalModule = await import("recharts");
    return {
        ...OriginalModule,
        ResponsiveContainer: ({ children }: { children: React.ReactElement }) => (
            <OriginalModule.ResponsiveContainer width={800} height={800}>
                {children}
            </OriginalModule.ResponsiveContainer>
        ),
    };
});


const mockPopulation = [
    { ano: '2021', populacao: 10030 },
    { ano: '2022', populacao: 8000 },
];
const mockNeighborhood = { id: 1, name: 'Neighborhood 1' };
const mockYears = '2021 - 2022';

describe('Chart', () => {
    beforeEach(() => {
        (usePopulation as unknown as Mock).mockReturnValue({
            selectedData: { population: mockPopulation, neighborhood: mockNeighborhood, years: mockYears },
        });
    });

    afterEach(() => {
        vi.clearAllMocks();
    });

    it('should render neighborhood name and years correctly', () => {
        const { container } = render(<Chart />);
        expect(container.firstChild).toMatchSnapshot();
    });

    it('should render BarChart with population data', () => {
        render(<Chart />);

        mockPopulation.forEach(({ populacao }) => {
            const bar = screen.getByText(populacao.toString());
            expect(bar).toBeInTheDocument();
        });
    });
});
