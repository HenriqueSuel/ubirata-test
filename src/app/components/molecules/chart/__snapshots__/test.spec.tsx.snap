// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Chart > should render neighborhood name and years correctly 1`] = `
<div
  class="bg-gray-800 px-6 py-6"
>
  <h1
    class="pb-8 text-3xl"
  >
    Neighborhood 1
     - 
    2021 - 2022
  </h1>
  <div
    class="recharts-responsive-container"
    style="width: 800px; height: 800px; min-width: 0;"
  >
    <div
      class="recharts-wrapper"
      role="region"
      style="position: relative; cursor: default; width: 800px; height: 800px;"
    >
      <svg
        class="recharts-surface"
        height="800"
        viewBox="0 0 800 800"
        width="800"
      >
        <title />
        <desc />
        <defs>
          <clippath
            id="recharts1-clip"
          >
            <rect
              height="760"
              width="730"
              x="65"
              y="5"
            />
          </clippath>
        </defs>
        <g
          class="recharts-layer recharts-bar"
        >
          <g
            class="recharts-layer recharts-bar-rectangles"
          >
            <g
              class="recharts-layer"
            >
              <g
                class="recharts-layer recharts-bar-rectangle"
              />
              <g
                class="recharts-layer recharts-bar-rectangle"
              />
            </g>
          </g>
        </g>
        <g
          class="recharts-layer recharts-cartesian-axis recharts-xAxis xAxis"
        >
          <line
            class="recharts-cartesian-axis-line"
            fill="none"
            height="30"
            orientation="bottom"
            stroke="#666"
            width="730"
            x="65"
            x1="65"
            x2="795"
            y="765"
            y1="765"
            y2="765"
          />
          <g
            class="recharts-cartesian-axis-ticks"
          >
            <g
              class="recharts-layer recharts-cartesian-axis-tick"
            >
              <line
                class="recharts-cartesian-axis-tick-line"
                fill="none"
                height="30"
                orientation="bottom"
                stroke="#666"
                width="730"
                x="65"
                x1="247.5"
                x2="247.5"
                y="765"
                y1="771"
                y2="765"
              />
              <text
                class="recharts-text recharts-cartesian-axis-tick-value"
                fill="#666"
                height="30"
                orientation="bottom"
                stroke="none"
                text-anchor="middle"
                width="730"
                x="247.5"
                y="773"
              >
                <tspan
                  dy="0.71em"
                  x="247.5"
                >
                  2021
                </tspan>
              </text>
            </g>
            <g
              class="recharts-layer recharts-cartesian-axis-tick"
            >
              <line
                class="recharts-cartesian-axis-tick-line"
                fill="none"
                height="30"
                orientation="bottom"
                stroke="#666"
                width="730"
                x="65"
                x1="612.5"
                x2="612.5"
                y="765"
                y1="771"
                y2="765"
              />
              <text
                class="recharts-text recharts-cartesian-axis-tick-value"
                fill="#666"
                height="30"
                orientation="bottom"
                stroke="none"
                text-anchor="middle"
                width="730"
                x="612.5"
                y="773"
              >
                <tspan
                  dy="0.71em"
                  x="612.5"
                >
                  2022
                </tspan>
              </text>
            </g>
          </g>
        </g>
        <g
          class="recharts-layer recharts-cartesian-axis recharts-yAxis yAxis"
        >
          <line
            class="recharts-cartesian-axis-line"
            fill="none"
            height="760"
            orientation="left"
            stroke="#666"
            width="60"
            x="5"
            x1="65"
            x2="65"
            y="5"
            y1="5"
            y2="765"
          />
          <g
            class="recharts-cartesian-axis-ticks"
          >
            <g
              class="recharts-layer recharts-cartesian-axis-tick"
            >
              <line
                class="recharts-cartesian-axis-tick-line"
                fill="none"
                height="760"
                orientation="left"
                stroke="#666"
                width="60"
                x="5"
                x1="59"
                x2="65"
                y="5"
                y1="765"
                y2="765"
              />
              <text
                class="recharts-text recharts-cartesian-axis-tick-value"
                fill="#666"
                height="760"
                orientation="left"
                stroke="none"
                text-anchor="end"
                width="60"
                x="57"
                y="765"
              >
                <tspan
                  dy="0.355em"
                  x="57"
                >
                  0
                </tspan>
              </text>
            </g>
            <g
              class="recharts-layer recharts-cartesian-axis-tick"
            >
              <line
                class="recharts-cartesian-axis-tick-line"
                fill="none"
                height="760"
                orientation="left"
                stroke="#666"
                width="60"
                x="5"
                x1="59"
                x2="65"
                y="5"
                y1="575"
                y2="575"
              />
              <text
                class="recharts-text recharts-cartesian-axis-tick-value"
                fill="#666"
                height="760"
                orientation="left"
                stroke="none"
                text-anchor="end"
                width="60"
                x="57"
                y="575"
              >
                <tspan
                  dy="0.355em"
                  x="57"
                >
                  3000
                </tspan>
              </text>
            </g>
            <g
              class="recharts-layer recharts-cartesian-axis-tick"
            >
              <line
                class="recharts-cartesian-axis-tick-line"
                fill="none"
                height="760"
                orientation="left"
                stroke="#666"
                width="60"
                x="5"
                x1="59"
                x2="65"
                y="5"
                y1="385"
                y2="385"
              />
              <text
                class="recharts-text recharts-cartesian-axis-tick-value"
                fill="#666"
                height="760"
                orientation="left"
                stroke="none"
                text-anchor="end"
                width="60"
                x="57"
                y="385"
              >
                <tspan
                  dy="0.355em"
                  x="57"
                >
                  6000
                </tspan>
              </text>
            </g>
            <g
              class="recharts-layer recharts-cartesian-axis-tick"
            >
              <line
                class="recharts-cartesian-axis-tick-line"
                fill="none"
                height="760"
                orientation="left"
                stroke="#666"
                width="60"
                x="5"
                x1="59"
                x2="65"
                y="5"
                y1="195"
                y2="195"
              />
              <text
                class="recharts-text recharts-cartesian-axis-tick-value"
                fill="#666"
                height="760"
                orientation="left"
                stroke="none"
                text-anchor="end"
                width="60"
                x="57"
                y="195"
              >
                <tspan
                  dy="0.355em"
                  x="57"
                >
                  9000
                </tspan>
              </text>
            </g>
            <g
              class="recharts-layer recharts-cartesian-axis-tick"
            >
              <line
                class="recharts-cartesian-axis-tick-line"
                fill="none"
                height="760"
                orientation="left"
                stroke="#666"
                width="60"
                x="5"
                x1="59"
                x2="65"
                y="5"
                y1="5"
                y2="5"
              />
              <text
                class="recharts-text recharts-cartesian-axis-tick-value"
                fill="#666"
                height="760"
                orientation="left"
                stroke="none"
                text-anchor="end"
                width="60"
                x="57"
                y="5"
              >
                <tspan
                  dy="0.355em"
                  x="57"
                >
                  12000
                </tspan>
              </text>
            </g>
          </g>
        </g>
      </svg>
      <div
        class="recharts-tooltip-wrapper"
        role="dialog"
        style="transform: translate(undefinedpx, undefinedpx); pointer-events: none; visibility: hidden; position: absolute; top: 0px; left: 0px;"
        tabindex="-1"
      >
        <div
          class="recharts-default-tooltip"
          style="margin: 0px; padding: 10px; background-color: rgb(255, 255, 255); border: 1px solid #ccc; white-space: nowrap;"
        >
          <p
            class="recharts-tooltip-label"
            style="margin: 0px;"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
